src/Raven.Client/Documents/Operations/Replication/PullReplicationDefinition.cs

```diff
@@ -1,23 +1,125 @@
+                [nameof(Disabled)] = Disabled,
+                [nameof(MentorNode)] = MentorNode,
+                [nameof(DelayReplicationFor)] = DelayReplicationFor,
+                [nameof(Filters)] = DynamicJsonValue.Convert(Filters),
+                [nameof(Mode)] = Mode
+            };
+        }
+
+        public void Validate(bool useSsl)
+        {
+            if (string.IsNullOrEmpty(Name))
+                throw new ArgumentNullException(nameof(Name));
+
+            if (useSsl == false)
+            {
+                if (Certificates?.Count > 0)
+                    throw new InvalidOperationException("Your server is unsecured and therefore you can't define pull replication with a certificate.");
+
+                return;
+            }
+
+            if (Certificates == null)
+                return;
+
+            if ((Filters?.Count ?? 0) == 0)
+                return;
+
+            if (Filters.Count != Certificates.Count)
+                throw new InvalidOperationException(
+                    "When using filtered replication, each certificate must have its own filter configuration, " +
+                    $"but found: {Certificates.Count:#,#;;0} certificates and only {Filters.Count:#,#;;0} filters");
+
+            foreach (KeyValuePair<string, string> kvp in Certificates.Where(kvp => Filters.ContainsKey(kvp.Key) == false))
+                throw new InvalidOperationException(
+                    "When using filtered replication, each certificate must have its own filter configuration, " +
+                    $"but there is no filter defined for certificate: {kvp.Key}");
+
+            foreach (KeyValuePair<string, FilteringOptions> kvp in Filters)
+            {
+                if ((kvp.Value?.AllowedPaths?.Length ?? 0) == 0)
+                    throw new InvalidOperationException($"Filter for {kvp} has a null or empty filter definition");
+
+                foreach (string path in kvp.Value.AllowedPaths)
+                {
+                    if (string.IsNullOrEmpty(path))
+                        throw new InvalidOperationException("Filtered replication AllowedPaths cannot have an empty / null filter");
+
+                    if (path[path.Length - 1] != '*')
+                        continue;
+
+                    if (path.Length > 1 && path[path.Length - 2] != '/' && path[path.Length - 2] != '-')
+                        throw new InvalidOperationException(
+                            $"When using '*' at the end of the allowed path, the previous character must be '/' or '-', but got: {path} for {kvp.Key}");
+                }
+            }
+        }
+
+        public class FilteringOptions : IDynamicJson
+        {
+            public string[] AllowedPaths;
+
+            public DynamicJsonValue ToJson()
+            {
+                return new DynamicJsonValue {[nameof(AllowedPaths)] = AllowedPaths};
+            }
+        }
+    }
+
+    public class RawPullReplicationDefinition
```

Comment:
Why we have this and `PullReplicationDefinition`  ?

Reply:
Mid way through refactoring things to avoid putting certificates inside the database record.